packages/core/admin/admin/src/core/apis/Widgets.ts

```diff
@@ -0,0 +1,57 @@
+/* eslint-disable check-file/filename-naming-convention */
+
+import invariant from 'invariant';
+import { To } from 'react-router-dom';
+
+import { Permission } from '../../../../shared/contracts/shared';
+
+import type { MessageDescriptor } from 'react-intl';
+
+type WidgetUID = `plugin::${string}.${string}` | `global::${string}`;
+
+interface WidgetArgs {
+  icon: React.ComponentType;
+  title: MessageDescriptor;
+  link?: {
+    label: MessageDescriptor;
+    href: To;
+  };
+  component: () => Promise<{ default: React.ComponentType }>;
+  pluginId?: string;
+  id: string;
+  permissions?: Permission[];
+}
+
+class Widgets {
+  widgets: Record<string, WidgetArgs>;
+
+  constructor() {
+    this.widgets = {};
+  }
+
+  register = (widget: WidgetArgs | WidgetArgs[]) => {
+    if (Array.isArray(widget)) {
+      widget.forEach((newWidget) => {
+        this.register(newWidget);
+      });
+    } else {
+      const { id, pluginId, component, title, icon } = widget;
+
+      invariant(id, 'An id must be provided');
+      invariant(component, 'A component must be provided');
+      invariant(title, 'A title must be provided');
+      invariant(icon, 'An icon must be provided');
+
+      const uid: WidgetUID = pluginId ? `plugin::${pluginId}.${id}` : `global::${id}`;
+
+      this.widgets[uid] = widget;
+    }
+  };
+
+  getAll = () => {
+    return this.widgets;
+  };
```

Comment:
I think it would be more convenient to have getAll return an array, so `Object.entries(this.widgets)`

If so then it would be helpful to do some transformations on the widgets object before storing it:
- having the uid as an object property would be neat, since otherwise we don't have it anymore via getAll in array form
- since id and pluginId are only used to generate the uid, I think we could not store them

Reply:
Makes sense, sounds good for the transformations I'll update

Reply:
> having the uid as an object property would be neat, since otherwise we don't have it anymore via getAll in array form

I'm doing this when storing

> since id and pluginId are only used to generate the uid, I think we could not store them

I thought it made more sense to do this transformation only in getAll, I thought [uid]: `{...widget, uid}` was kind of strange. But now I'm thinking it makes more sense to do all the transformations in one place. wdyt?

Reply:
I decided to do this: But now I'm thinking it makes more sense to do all the transformations in one place

which I believe is what you were suggesting anyway 🙃